### PR TITLE
Add support to custom bin edges to histo and profile models

### PR DESCRIPTION
**Pull request by dpiparo**

### CODE REVIEW

File: tree/treeplayer/inc/ROOT/TDFHistoModels.hxx [hunk 1]
```diff
@@ -30,12 +30,15 @@ struct TH1DModel {
...
    double fXLow = 0.;
    double fXUp = 64.;
+   std::vector<double> fBinXEdges;
 
    TH1DModel() = default;
    TH1DModel(const TH1DModel &) = default;
    ~TH1DModel();
    TH1DModel(const ::TH1D &h);
    TH1DModel(const char *name, const char *title, int nbinsx, double xlow, double xup);
+   TH1DModel(const char *name, const char *title, Int_t nbinsx, const float *xbins);
```
Comment (bluehood) [inline tree/treeplayer/inc/ROOT/TDFHistoModels.hxx]: Any specific reason why `nbinsx` is `Int_t` here but `int` above?

File: tree/treeplayer/inc/ROOT/TDFInterface.hxx [hunk 2]
```diff
@@ -511,14 +510,17 @@ public:
    ////////////////////////////////////////////////////////////////////////////
    /// \brief Creates a custom column with a value dependent on the processing slot and the current entry.
    /// \param[in] name The name of the custom column.
-   /// \param[in] expression Function, lambda expression, functor class or any other callable object producing the temporary value. Returns the value that will be assigned to the custom column.
+   /// \param[in] expression Function, lambda expression, functor class or any other callable object producing the
+   /// temporary value. Returns the value that will be assigned to the custom column.
    /// \param[in] columns Names of the columns/branches in input to the producer function (excluding slot and entry).
```
Comment (bluehood) [inline tree/treeplayer/inc/ROOT/TDFInterface.hxx]: This is clang-format messing up our docs

File: tree/treeplayer/inc/ROOT/TDFInterface.hxx [hunk 3]
```diff
@@ -511,14 +510,17 @@ public:
...
    ///
    /// This alternative implementation of `Define` is meant as a helper in writing entry-specific, thread-safe custom
    /// columns. The expression must be a callable of signature R(unsigned int, ULong64_t, T1, T2, ...) where `T1, T2...`
    /// are the types of the columns that the expression takes as input. The first parameter is reserved for an unsigned
-   /// integer representing a "slot number". TDataFrame guarantees that different threads will invoke the expression with
-   /// different slot numbers - slot numbers will range from zero to ROOT::GetImplicitMTPoolSize()-1. The second parameter
+   /// integer representing a "slot number". TDataFrame guarantees that different threads will invoke the expression
+   /// with
+   /// different slot numbers - slot numbers will range from zero to ROOT::GetImplicitMTPoolSize()-1. The second
+   /// parameter
```
Comment (bluehood) [inline tree/treeplayer/inc/ROOT/TDFInterface.hxx]: again wrong clang-formatting I think

File: tree/treeplayer/inc/ROOT/TDFInterface.hxx [hunk 4]
```diff
@@ -1444,7 +1484,8 @@ public:
    /// \brief Return the sum of processed column values (*lazy action*)
    /// \tparam T The type of the branch/column.
    /// \param[in] columnName The name of the branch/column.
-   /// \param[in] initValue Optional initial value for the sum. If not present, the column values must be default-constructible.
+   /// \param[in] initValue Optional initial value for the sum. If not present, the column values must be
+   /// default-constructible.
```
Comment (bluehood) [inline tree/treeplayer/inc/ROOT/TDFInterface.hxx]: clang-format mistake

**Comment by bluehood** [comment]
Jenkins errors are false positives and the PR looks good to me, except for a few formalities.

File: tree/treeplayer/inc/ROOT/TDFHistoModels.hxx (hunk 1 above)
Comment (dpiparo) [reply tree/treeplayer/inc/ROOT/TDFHistoModels.hxx]: No. It was a mistake in the first implementation. Already fixed on my local branch :)

File: tree/treeplayer/inc/ROOT/TDFInterface.hxx (hunk 2 above)
Comment (dpiparo) [reply tree/treeplayer/inc/ROOT/TDFInterface.hxx]: I thought I protected all code parts which need to be left alone by clang-format. I will take care.

**Comment by dpiparo** [comment]
On top of what we have already and a bit more of pyroot testing, I need to figure out what to do with the signatures. The fact that the size of bins is an integer and not a size_t is a problem for initialiser lists. Now one would do in Python:
```
v = ROOT.std.vector('double')()
for i in [0,1,2,3,4,5]: v.push_back(i)
tdf.Histo1D(('myhisto', 'myhisto', int(v.size()), v.data()),"mycol")
```
that cast is bad. What would be desirable would be:
```
tdf.Histo1D(('myhisto', 'myhisto', v.size(), v.data()),"mycol")
```
or even
```
tdf.Histo1D(('myhisto', 'myhisto', v),"mycol")
```
The last but one line involves a change in the interface (which makes it incompatible with the one of TH*) while the last one also a change in PyROOT for implicit conversions to array_view<T>.

**Comment by bluehood** [comment]
@dpiparo I agree the cast is not nice, but isn't that issue separate from the one this PR addresses?
If I understand correctly pyROOT users have always been required to perform that cast due to the C++ signatures of `TH*`.

**Comment by dpiparo** [comment]
Yes, it is of course separate but when developing this new feature it became particularly striking to me, at least for this particular case.

**Pull request merged**